### PR TITLE
[YUNIKORN-869] Improve coverage of node_collection.go

### PR DESCRIPTION
**Pull request by 0yukali0**
### What is this PR for?
Add some case to cover these functions
1. GetIterator
2. GetNodeSortingPolicy
3. SetSodeSortingPolicy

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-869

### How should this be tested?
Use "make test" command in yonikorn repo and then use "go tool cover --func=coverage.txt" to check the result

### Screenshots (if appropriate)
截圖 2022-04-13 下午7 28 29
### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

File: pkg/scheduler/objects/node_collection_test.go [hunk 1]
```diff
@@ -120,3 +121,56 @@ func TestNodeCollection_GetNodes(t *testing.T) {
...
+	}{
+		{"fair", "fair"},
+		{"bin", "bin"},
+	}
+
+	for _, tt := range tests {
+		testname := fmt.Sprintf("Get method of base_collection:%s %s", tt.input, tt.except)
+		t.Run(testname, func(t *testing.T) {
+			policy := NewNodeSortingPolicy(tt.input, weights)
+			nc.SetNodeSortingPolicy(policy)
```
Comment (wilfred-s) [inline pkg/scheduler/objects/node_collection_test.go:144]: Changing the policy changes the node ordering. Should this not be part of the test?
* nil sorting returns node in ??random?? order (not sure)
* fair sorting will return least loaded node first
* bin sorting should return most loaded node first.

We should have thus at least two nodes and make sure that the order they come back in is correct after the update.

**Review by wilfred-s — CHANGES_REQUESTED** [review]
Test are looking promising but they leave a number of checks out to make sure we correctly test things.

File: pkg/scheduler/objects/node_collection_test.go [hunk 2]
```diff
@@ -120,3 +121,56 @@ func TestNodeCollection_GetNodes(t *testing.T) {
 	nodes = nc.GetNodes()
 	assert.Equal(t, 1, len(nodes), "list is missing node")
 }
+
+func TestGetNodeSortingPolicy(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (0yukali0) [reply pkg/scheduler/objects/node_collection_test.go:125]: Ok, I will add it.

File: pkg/scheduler/objects/node_collection_test.go [hunk 3]
```diff
@@ -120,3 +121,56 @@ func TestNodeCollection_GetNodes(t *testing.T) {
...
+
+	bc := initBaseCollection()
+	var nc NodeCollection = bc
+	// Register callback listener
+	node.AddListener(bc)
+	if err := nc.AddNode(node); err != nil {
+		t.Errorf("Adding a node should be worked.")
+	}
+	// Callback trigger
+	if err := node.Reserve(app, ask); err != nil {
```
[the remark this answers is not in the document]
Comment (0yukali0) [reply pkg/scheduler/objects/node_collection_test.go:169]: Got it.

File: pkg/scheduler/objects/node_collection_test.go (hunk 1 above)
Comment (0yukali0) [reply pkg/scheduler/objects/node_collection_test.go:144]: Got it.
I will add these scenarios in SetSortingPolicy and GetNodeIterator unit test.
Thanks for your feedback, @wilfred-s

**Comment by 0yukali0** [comment]
Hi @wilfred-s , I updated related work to cover following problems.

1.  Get nc from new function and then test the node order when policy is nil. 
2.  Set different policy including fair and binpacking  to test the node order.
3.  Set fair as default policy when policy is unknown.

**Review by wilfred-s — APPROVED** [review]
LGTM

**Pull request closed without merging**